File: Java/shakerSort.java
public class ShakerSort {

    // Method to perform Shaker Sort
    public static void shakerSort(int[] array) {
        int n = array.length; // Get the length of the array
        boolean swapped; // Flag to check if a swap has occurred
        int start = 0; // Starting index
        int end = n - 1; // Ending index

        // Repeat until the entire array is sorted
        do {
            swapped = false;

            // Pass from left to right
            for (int i = start; i < end; i++) {
                // If the current element is greater than the next element
                if (array[i] > array[i + 1]) {
                    // Swap the elements
                    int temp = array[i];
                    array[i] = array[i + 1];
                    array[i + 1] = temp;
                    swapped = true; // Set swapped to true
                }
            }

            // If no swaps occurred, the array is sorted
            if (!swapped) {
                break;
            }

            // Reduce the end index for the next pass
            end--;

            swapped = false; // Reset the flag for the next pass

            // Pass from right to left
            for (int i = end; i > start; i--) {
                // If the current element is less than the previous element
                if (array[i] < array[i - 1]) {
                    // Swap the elements
                    int temp = array[i];
                    array[i] = array[i - 1];
                    array[i - 1] = temp;
                    swapped = true; // Set swapped to true
                }
            }

            // Increase the start index for the next pass
            start++;

        } while (swapped); // Continue until no swaps occur
    }

    // Method to print the array
    public static void printArray(int[] array) {
        // Iterate through the array and print each element
        for (int value : array) {
            System.out.print(value + " ");
        }
        System.out.println();
    }

    // Main method to test the sorting algorithm
    public static void main(String[] args) {
        // Sample array to be sorted
        int[] array = {64, 34, 25, 12, 22, 11, 90};

        System.out.println("Original array:");
        printArray(array); // Print the original array

        // Call the shaker sort method
        shakerSort(array);

        System.out.println("Sorted array:");
        printArray(array); // Print the sorted array
    }
}

File: Go/shakerSort.go
package main

import (
	"fmt"
)

// shakerSort sorts an array using the shaker sort algorithm.
func shakerSort(arr []int) {
	n := len(arr)      // Get the length of the array
	start := 0         // Starting index for the left to right pass
	end := n - 1       // Ending index for the right to left pass
	swapped := true     // Flag to check if a swap occurred

	for swapped {      // Continue until no swaps are made
		swapped = false // Reset the swapped flag for this pass

		// Left to right pass
		for i := start; i < end; i++ {
			// If the current element is greater than the next, swap them
			if arr[i] > arr[i+1] {
				arr[i], arr[i+1] = arr[i+1], arr[i] // Swap elements
				swapped = true                      // Mark that a swap occurred
			}
		}

		// If no swaps were made, the array is sorted
		if !swapped {
			break // Exit if the array is already sorted
		}

		// Decrease the end index since the last element is sorted
		end--

		swapped = false // Reset the swapped flag for the next pass

		// Right to left pass
		for i := end; i > start; i-- {
			// If the current element is less than the previous, swap them
			if arr[i] < arr[i-1] {
				arr[i], arr[i-1] = arr[i-1], arr[i] // Swap elements
				swapped = true                       // Mark that a swap occurred
			}
		}

		// Increase the start index since the first element is sorted
		start++
	}
}

// main function to test the shakerSort
func main() {
	arr := []int{64, 34, 25, 12, 22, 11, 90} // Sample array to be sorted
	fmt.Println("Original array:", arr)

	shakerSort(arr) // Call the sorting function

	fmt.Println("Sorted array:", arr) // Print the sorted array
}
